core/species.cpp

```diff
@@ -3854,126 +4026,405 @@ void Species::SimplifyTreeSequence(void)
-	
-	// and reset our elapsed time since last simplification, for auto-simplification
-	simplify_elapsed_ = 0;
-	
-	// as a side effect of simplification, update a "model has coalesced" flag that the user can consult, if requested
-	if (running_coalescence_checks_)
-		CheckCoalescenceAfterSimplification();
+	// note that we leave things in a partially completed state; the node table still needs to be filtered!
 }
 
-void Species::CheckCoalescenceAfterSimplification(void)
+void Species::SimplifyAllTreeSequences(void)
 {
 #if DEBUG
-	if (!recording_tree_ || !running_coalescence_checks_)
-		EIDOS_TERMINATION << "ERROR (Species::CheckCoalescenceAfterSimplification): (internal error) coalescence check called with recording or checking off." << EidosTerminate();
+	if (!recording_tree_)
+		EIDOS_TERMINATION << "ERROR (Species::SimplifyAllTreeSequences): (internal error) tree sequence recording method called with recording off." << EidosTerminate();
 #endif
 	
-	// Copy the table collection; Jerome says this is unnecessary since tsk_table_collection_build_index()
-	// does not modify the core information in the table collection, but just adds some separate indices.
-	// However, we also need to add a population table, so really it is best to make a copy I think.
-	tsk_table_collection_t tables_copy;
-	int ret;
-	
-	ret = tsk_table_collection_copy(&tables_, &tables_copy, 0);
-	if (ret < 0) handle_error("tsk_table_collection_copy", ret);
-	
-	// Our tables copy needs to have a population table now, since this is required to build a tree sequence
-	WritePopulationTable(&tables_copy);
-	
-	ret = tsk_table_collection_build_index(&tables_copy, 0);
-	if (ret < 0) handle_error("tsk_table_collection_build_index", ret);
-	
-	tsk_treeseq_t ts;
-	
-	ret = tsk_treeseq_init(&ts, &tables_copy, 0);
-	if (ret < 0) handle_error("tsk_treeseq_init", ret);
+	// if we have no recorded nodes, there is nothing to simplify; note that the nodes table is shared
+	if (treeseq_[0].tables_.nodes.num_rows == 0)
+		return;
 	
-	// Collect a vector of all extant haplosome node IDs
-	int haplosome_count_per_individual = HaplosomeCountPerIndividual();
-	std::vector<tsk_id_t> all_extant_nodes;
+	std::vector<tsk_id_t> samples;
 	
-	for (auto subpop_iter : population_.subpops_)
+	// BCH 7/27/2019: We now build a hash table containing all of the entries of remembered_nodes_,
+	// so that the find() operations in the loop below can be done in constant time instead of O(N) time.
+	// We need to be able to find out the index of an entry, in remembered_nodes_, once we have found it;
+	// that is what the mapped value provides, whereas the key value is the tsk_id_t we need to find below.
+	// We do all this inside a block so the map gets deallocated as soon as possible, to minimize footprint.
+	// BCH 12/9/2024: The point of all this kerfuffle is that an extant individual might also be a
+	// remembered individual, and we don't want to put it into the samples vector twice, I think.
```

Comment:
Maybe also that we need to put the samples first?

Reply:
I changed the comment to:

```
// BCH 12/9/2024: The point of all this kerfuffle with the lookup table is that an extant individual
// might also be a remembered individual, and we don't want to put it into the samples vector twice,
// I think; otherwise we could just throw the remembered nodes and extant individuals into `samples`
// with no lookup table complication.
```

I was trying to figure out, and then explain to myself, why `remembered_nodes_lookup` is needed, since that chunk of code is surprisingly complicated.  Hopefully my revised comment is clearer, and accommodates your feedback here.  Resolving this thread.